### PR TITLE
[GSoC] Simplified UpdateNote asynctask callbacks

### PR DESCRIPTION
**Pull request by divyansh-dxn**
## Purpose / Description
When updating a single card, it is not required to show continuous progress since it a short task and user can be notified after the task has been finished. So instead of using onProgressUpdate callback, it is good to use onPostExecute for UI updates after the card has been updated.

THIS PR WILL HELP MIGRATING UPDATENOTES TO COROUTINE

## How this has been tested?
Passes existing tests and runs as expected on Realme 6i API 30 and Pixel 4a AVD API 31

## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt [hunk 1]
```diff
@@ -398,50 +398,43 @@ abstract class AbstractFlashcardViewer :
...
-                return
-            }
-            onCardEdited(mCurrentCard)
-            if (sDisplayAnswer) {
-                mSoundPlayer.resetSounds() // load sounds from scratch, to expose any edit changes
-                mAnswerSoundsAdded = false // causes answer sounds to be reloaded
-                generateQuestionSoundList() // questions must be intentionally regenerated
-                displayCardAnswer()
+        override fun onPostExecute(result: Card?) {
+            if (result != null) {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt:407]: You could reduce the indentation by returning early:
```

if (result == null) {
    closeReviewer(DeckPicker.RESULT_DB_ERROR, false)
    return
}
if (mCurrentCard !== result) {
```

to actually use the return statement which is currently redundant.

**Review by lukstbit — CHANGES_REQUESTED** [review]
LGTM with two styles issues.

**Review by lukstbit — APPROVED** [review]

**Comment by divyansh-dxn** [comment]
I like the simplification that is going on. Previously the showing and hiding of progress bar and handling error in CardBrowser is extracted into some base-classes, while going through the code it as hard to follow up and navigating class to class. TBH, some code repetition in more understandable way is better that code reuse which is hard to understand. Event these components were not too long to be extracted out even.

This is my opinion. Also when migrating to coroutines, progress will be handled by withProgress{} and error with launchCatchingtask{}, amazing wrappers implemented by @dae.

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt [hunk 2]
```diff
@@ -398,50 +398,43 @@ abstract class AbstractFlashcardViewer :
...
+                     * then we need to display it as a new card, without showing the answer.
+                     */
+                    sDisplayAnswer = false
+                }
+                currentCard = result
+                TaskManager.launchCollectionTask(PreloadNextCard()) // Tasks should always be launched from GUI. So in
+                // listener and not in background
+                if (mCurrentCard == null) {
+                    // If the card is null means that there are no more cards scheduled for review.
+                    showProgressBar()
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt]: I believe you can remove the `showProgressBar`. It was (theoretically) useful if we assumed some time may take until we end the computation. But now it makes no sense anymore

**Review by Arthur-Milchior — APPROVED** [review]
I agree it's a change that does not change the behavior*, so I'm fine merging it. One remark, but that can wait for the time where this task is transformed into a coroutine

* at least in the case where there is no exception
